lib/ruby_jard/session.rb

```diff
@@ -11,13 +13,32 @@ module RubyJard
   # other processes. Therefore, an internal, jard-specific data mapping should
   # be built.
   class Session
-    attr_accessor :threads, :current_frame, :current_backtrace
+    class << self
+      extend Forwardable
+
+      def_delegators :instance,
+                     :attach, :lock, :update, :flush,
+                     :threads, :current_frame, :current_backtrace,
+                     :output_buffer, :append_output_buffer,
+                     :secondary_output_buffer, :append_secondary_output_buffer, :flush_secondary_output_buffer
+
+      def instance
+        @instance ||= new
+      end
+    end
+
+    OUTPUT_BUFFER_LENGTH = 10_000 # 10k lines
+
+    attr_accessor :threads, :current_frame, :current_backtrace, :output_buffer
 
     def initialize(options = {})
       @options = options
       @started = false
       @session_lock = Mutex.new
 
+      @output_buffer = []
```

Comment:
Maybe you can use a `StringIO` to keep the interface the same as Tempfile?

Reply:
I am to implement a ring buffer, limit the cached output of around 10k lines. StringIO is not very suitable for this purpose. Besides, when doing micro-benchmarking, StringIO is around 20% slower than pure array implementation, and the differences keep growing as the stress test increases.

Reply:
Hmm, interesting. I did a simple benchmark for naive array implementation vs StringIO and the result seems to be not bad:

```
require 'benchmark'

pager = STDOUT
string = 'longlonglonglonglonglonglonglonglong'

result = Benchmark.bmbm do |x|
  x.report('StringIO') do
    io = StringIO.new
    10000.times { io.puts string }
    pager.write(io.string)
  end

  x.report('StringIO') do
    io = []
    10000.times { io << string }
    
    io.each do |item|
      pager.write(item)
    end
  end
end

puts result

# =>   StringIO 0.003346   0.001859   0.005205 (  0.025737)
#      Array    0.012189   0.007254   0.019443 (  0.042336)
```

Reply:
Ah. My benchmark is slightly different, but it's use your code, and increase size a little bit (like the sample rate I did in the description).

```ruby
require 'stringio'
require 'benchmark'

pager = STDOUT
string = 'I, [2020-08-18T23:03:33.158859 #575832]  INFO -- : Started GET "/calendars/2?token=123456" for 127.0.0.1 at 2020-08-18 23:03:33 +0700'

result = Benchmark.bmbm do |x|
  x.report('StringIO') do
    io = StringIO.new
    500_000.times { io.puts string }
    pager.write(io.string)
  end

  x.report('Array') do
    io = []
    500_000.times { io << string }

    io.each do |item|
      pager.write(item)
    end
  end
end

puts result
```

The result is 

```
               user     system      total        real
StringIO   0.081003   0.785662   0.866665 ( 9.251325)
Array      0.498736   0.836263   1.334999 (  8.020491)
```

I usually use real column result. Not sure whether it's correct or not. In my case, 12.5% faster. However, it is just raw benchmark. If adding my constraint (limit latest 10k lines), I couldn't find an easy way to do it with StringIO though :thinking: